### ISSUE TITLE
Custom Kernel using LSTM network doesn't work

### ISSUE THREAD

**Issue by ya54** [GPflow/GPflow#1689]
I created a custom kernel using LSTM:
GPFlow is not accepting this network and keeps throwing exceptions and errors. 
The input data is daily temperature a sequence of numbers. 
It never worked and keeps complaining. 
LSTM is very famous along with it's versions like bidirectional LSTM, GRU, etc... but yet, I'm struggling to get it working.
Can you please check why this is an issue and provide a working example with LSTM?
The code is:

import warnings
warnings.filterwarnings("ignore")
import os
os.environ['TF_CPP_MIN_LOG_LEVEL'] = '3' 

try:
    tf.compat.v1.logging.set_verbosity(tf.compat.v1.logging.ERROR)
except Exception as e:
    print('')
    
    
import numpy as np
import pandas as pd
import matplotlib.pylab as plt
from statsmodels.tsa.stattools import adfuller
from statsmodels.tsa.seasonal import seasonal_decompose
from statsmodels.tsa.arima_model import ARIMA
#from pmdarima.arima import auto_arima
from sklearn.metrics import mean_squared_error, mean_absolute_error
import math
import tensorflow as tf
from tensorflow import keras
import seaborn as sns
from pylab import rcParams
from matplotlib import rc

df = pd.read_csv('daily-min-temperatures.csv')

data = df.set_index(['Date'])
data.head()

def adfuller_test(series, signif=0.05, name='', verbose=False):
    """Perform ADFuller to test for Stationarity of given series and print report"""
    r = adfuller(series, autolag='AIC')
    output = {'test_statistic':round(r[0], 4), 'pvalue':round(r[1], 4), 'n_lags':round(r[2], 4), 'n_obs':r[3]}
    p_value = output['pvalue'] 
    def adjust(val, length= 6): return str(val).ljust(length)

    # Print Summary
    print(f'    Augmented Dickey-Fuller Test on "{name}"', "\n   ", '-'*47)
    print(f' Null Hypothesis: Data has unit root. Non-Stationary.')
    print(f' Significance Level    = {signif}')
    print(f' Test Statistic        = {output["test_statistic"]}')
    print(f' No. Lags Chosen       = {output["n_lags"]}')

    for key,val in r[4].items():
        print(f' Critical value {adjust(key)} = {round(val, 3)}')

    if p_value <= signif:
        print(f" => P-Value = {p_value}. Rejecting Null Hypothesis.")
        print(f" => Series is Stationary.")
    else:
        print(f" => P-Value = {p_value}. Weak evidence to reject the Null Hypothesis.")
        print(f" => Series is Non-Stationary.")   

adfuller_test(data['Temp'], name='Temp')
print('\n')

data['date'] = pd.to_datetime(data.index)
data['year'] = data['date'].dt.year
data['month'] = data['date'].dt.month
data['day'] = data['date'].dt.day

train_size = int(len(data) * 0.8)
test_size = len(data) - train_size
train, test = data[0:train_size], data.iloc[train_size:len(data)]
print(len(train), len(test))

from numpy import array
# split a univariate sequence into samples
def split_sequence(sequence, n_steps):
    X, y = list(), list()
    for i in range(len(sequence)):
        # find the end of this pattern
        end_ix = i + n_steps
        # check if we are beyond the sequence
        if end_ix > len(sequence)-1:
            break
        # gather input and output parts of the pattern
        seq_x, seq_y = sequence[i:end_ix], sequence[end_ix]
        X.append(seq_x)
        y.append(seq_y)
    return array(X), array(y)

# choose a number of time steps
time_steps = 10
# split into samples
train_X,train_y = split_sequence(train['Temp'], time_steps)
test_X,test_y = split_sequence(test['Temp'],time_steps)
#print(train_X[0:5])
#print(test_X[0:5])
train_X.shape

train_y= train_y.reshape(-1,1)

train_X.shape

import gpflow
from gpflow.utilities import print_summary

class kernelwithLSTM(gpflow.kernels.Kernel):
    def __init__(self,input_size):
        super().__init__(active_dims=[0])
        self.variance = gpflow.Parameter(1.0)
        self.input_size = input_size
        self.lstm = tf.keras.Sequential(
                [
                    tf.keras.layers.InputLayer(input_shape=self.input_size),
                    tf.keras.layers.Reshape((-1,1)),                    
                    tf.keras.layers.LSTM(units=64, return_sequences=True, activation='tanh',dtype='float64'),
                    tf.keras.layers.Dense(units=1,dtype='float64'), 
                    
                ]
            )

        self.lstm.build()
  
    def K(self, a_input, b_input=None):
        if b_input is None:
            b_input = a_input
        #print('a_input',a_input)
        #print('b_input',b_input)
        transformed_a = self.lstm(a_input) 
        transformed_b = self.lstm(b_input)
        return self.variance * tf.minimum(transformed_a, tf.transpose(transformed_b))  # this returns a 2D tensor

    def K_diag(self, a_input):
        #print('k_diag_a',a_input)
        transformed_a = self.lstm(a_input)
        return self.variance * tf.reshape(transformed_a, (-1,))  # this returns a 1D tensor

print('train_X.shape:', train_X.shape)
input_size= train_X.shape[0]
k_brownian = kernelwithLSTM(input_size)
m = gpflow.models.GPR((train_X, train_y), kernel=k_brownian)

opt = gpflow.optimizers.Scipy()
opt.minimize(m.training_loss, variables=m.trainable_variables)
mean, var = m.predict_y(test_X)

print_summary(m)

stdev = tf.math.sqrt(var)

# For calculating range, I have used z ~= 3 in the formula (mean-z*stdev, mean+3*stdev) 
lower_bounds = mean - 3 * stdev
upper_bounds = mean + 3 * stdev

lb= tf.reshape(lower_bounds,[-1])
ub =tf.reshape(upper_bounds,[-1])
x= np.arange(720)

plt.figure(figsize=(15,7))
plt.plot(test_y,label=' Actual values',marker='.',linewidth=2)
plt.plot(mean,label='Predicted values',marker='.',linewidth=1.7,color='coral')
plt.plot(lower_bounds,linewidth=0.1,color='grey')
plt.plot(upper_bounds,linewidth=0.1,color='grey')
plt.fill_between(x,lb,ub,color='lightgrey',alpha=0.5)
plt.title('Gaussian Process Regression with LSTM')
plt.xlabel('Date')
plt.ylabel('Temperature')
plt.legend(loc='best')
plt.show()

# Computing error
from math import sqrt
from sklearn.metrics import mean_squared_error, mean_absolute_error

RMSE = sqrt(mean_squared_error(test_y,mean))
RMSE

![gpflow_lstm_kernel_error](https://user-images.githubusercontent.com/30779626/119662814-cdae6280-bdff-11eb-9dfc-60181e35e2cc.png)

**Comment by ya54** [comment]
duplicate of https://github.com/GPflow/GPflow/issues/1692
